## README

Annotate potential novel peptide/proteins from PacBio long-read RNA seq  
Infer the translation trap/stall mechanism

Input:
- Mapped BAM file after quality control
- Annotated genome gff and FASTA file

Output:
- Excel file: RNA Isoform annotation, Novel ORF annotation
- FASTA file: protein seq

FASTA file will be used in mass spectrometry based proteomics analysis.

## Locate Abnormal Transcripts

Label per isoform against the known gene annotation.

### All isoforms

In [1]:
# ============================================================
# Isoform clustering + annotation against genes
# Regenerated version:
#   - NO QC filtering here (use your pre-filtered BAM directly)
#   - FIXED coordinate bug for minus-strand isoforms
#   - stores both transcript-oriented 5'/3' and genomic start/end
#
# Input:
#   - high-confidence BAM
#   - genome GFF/GFF3
#
# Output:
#   - isoform_clusters_annotated.tsv
#   - isoform_clusters_annotated.xlsx
#   - optional isoform_membership.tsv
# ============================================================

from __future__ import annotations
import os
from pathlib import Path
from typing import Optional, Tuple, List, Dict

import numpy as np
import pandas as pd
import pysam


# ----------------------------
# Inputs
# ----------------------------
MOTHER_FOLDER = ".."   # change if needed
BAM_PATH = MOTHER_FOLDER + "/syn1.PacBio.FLNC.sorted.HQ.bam"   # change if needed
GFF_PATH = MOTHER_FOLDER + "/Genomes_Input/syn1.genes.gff3"                   # change if needed

OUT_FOLDER = MOTHER_FOLDER + "/isoform_annotation"
Path(OUT_FOLDER).mkdir(parents=True, exist_ok=True)

OUT_ISOFORMS_TSV   = OUT_FOLDER + "/isoform_clusters_annotated.tsv"
OUT_ISOFORMS_XLSX  = OUT_FOLDER + "/isoform_clusters_annotated.xlsx"
OUT_MEMBERSHIP_TSV = OUT_FOLDER + "/isoform_membership.tsv"   # set to None to disable

In [ ]:
# ----------------------------
# BAM handling
# ----------------------------
REQUIRE_PRIMARY = True   # keep only primary alignments


# ----------------------------
# Isoform clustering parameters
# ----------------------------
# grouping is by transcript-oriented 5' and 3' ends
BIN5 = 20
BIN3 = 30

MIN_ISOFORM_READS = 2

DO_POST_MERGE = True
MERGE_5P_TOL_BP = 15
MERGE_3P_TOL_BP = 15


# ----------------------------
# Gene / annotation rules
# ----------------------------
GENE_FEATURE_TYPES = {"gene", "CDS"}   # prefer gene, fallback CDS
GENE_ID_KEYS       = ["locus_tag", "ID", "Name", "gene", "gene_id"]
GENE_NAME_KEYS     = ["gene", "Name", "product", "locus_tag", "ID"]

MIN_GENE_OVERLAP_BP = 50
PURE_FRAC = 0.95


# ----------------------------
# Helpers
# ----------------------------
def overlap_len(a0: int, a1: int, b0: int, b1: int) -> int:
	return max(0, min(a1, b1) - max(a0, b0))

def mad(x: np.ndarray) -> float:
	if x.size <= 1:
		return 0.0
	m = np.median(x)
	return float(np.median(np.abs(x - m)))

def bin_pos(pos0: int, binsz: int) -> int:
	return int((pos0 // binsz) * binsz)

def aligned_interval_and_ends(read: pysam.AlignedSegment) -> Optional[Dict[str, int]]:
	"""
	Returns a dict with:
	  strand  : '+' or '-'
	  g_start0: genomic interval start (0-based, inclusive)
	  g_end0  : genomic interval end   (0-based, exclusive)
	  pos5p0  : transcript-oriented 5' aligned end
	  pos3p0  : transcript-oriented 3' aligned end

	Softclips are excluded because pysam reference_start/reference_end
	are aligned coordinates.
	"""
	r0 = read.reference_start
	r1 = read.reference_end
	if r0 is None or r1 is None or r1 <= r0:
		return None

	strand = "-" if read.is_reverse else "+"

	if strand == "+":
		pos5p0 = int(r0)
		pos3p0 = int(r1)
	else:
		pos5p0 = int(r1)
		pos3p0 = int(r0)

	return {
		"strand": strand,
		"g_start0": int(r0),
		"g_end0": int(r1),
		"pos5p0": pos5p0,
		"pos3p0": pos3p0,
	}

def parse_gff_attributes(attr_str: str) -> Dict[str, str]:
	d = {}
	if pd.isna(attr_str):
		return d
	for field in str(attr_str).split(";"):
		field = field.strip()
		if not field:
			continue
		if "=" in field:
			k, v = field.split("=", 1)
		elif " " in field:
			k, v = field.split(" ", 1)
			v = v.strip('"')
		else:
			continue
		d[k.strip()] = v.strip()
	return d

def first_attr(attrs: Dict[str, str], keys: List[str], default: str = "") -> str:
	for k in keys:
		if k in attrs and attrs[k] not in ("", ".", None):
			return str(attrs[k])
	return default

def load_genes_from_gff(gff_path: str,
						feature_types: set[str] = {"gene", "CDS"}) -> pd.DataFrame:
	rows = []
	with open(gff_path, "r") as fh:
		for line in fh:
			if not line.strip() or line.startswith("#"):
				continue
			parts = line.rstrip("\n").split("\t")
			if len(parts) != 9:
				continue
			chrom, source, feature, start1, end1, score, strand, phase, attrs = parts
			if feature not in feature_types:
				continue
			if strand not in {"+", "-"}:
				continue

			try:
				start1 = int(start1)
				end1   = int(end1)
			except ValueError:
				continue

			if end1 < start1:
				continue

			attrd = parse_gff_attributes(attrs)
			gene_id   = first_attr(attrd, GENE_ID_KEYS, default="")
			gene_name = first_attr(attrd, GENE_NAME_KEYS, default=gene_id)

			rows.append({
				"chrom": chrom,
				"start0": start1 - 1,  # GFF 1-based inclusive -> 0-based half-open
				"end0": end1,
				"strand": strand,
				"gene_id": gene_id if gene_id else f"{feature}:{chrom}:{start1}-{end1}:{strand}",
				"gene_name": gene_name if gene_name else gene_id,
				"feature_type": feature,
			})

	df = pd.DataFrame(rows)
	if df.empty:
		raise ValueError("No gene/CDS features parsed from GFF.")

	if (df["feature_type"] == "gene").any():
		df = df[df["feature_type"] == "gene"].copy()
	else:
		df = df[df["feature_type"] == "CDS"].copy()

	df = df.sort_values(["chrom", "start0", "end0", "strand"]).reset_index(drop=True)
	return df

def post_merge_isoforms(df: pd.DataFrame,
						tol5: int = 15,
						tol3: int = 15) -> pd.DataFrame:
	"""
	Merge isoforms within each (chrom, strand) if transcript-oriented
	5' and 3' ends are close.
	"""
	if df.empty:
		return df.copy()

	out_rows = []

	for (chrom, strand), g in df.sort_values(
		["chrom", "strand", "pos5p0", "pos3p0"]
	).groupby(["chrom", "strand"], sort=False):

		current = None

		for row in g.to_dict("records"):
			if current is None:
				current = row.copy()
				continue

			same_5 = abs(int(row["pos5p0"]) - int(current["pos5p0"])) <= tol5
			same_3 = abs(int(row["pos3p0"]) - int(current["pos3p0"])) <= tol3

			if same_5 and same_3:
				n1 = int(current["n_reads"])
				n2 = int(row["n_reads"])
				n  = n1 + n2

				current["pos5p0"] = int(round((current["pos5p0"] * n1 + row["pos5p0"] * n2) / n))
				current["pos3p0"] = int(round((current["pos3p0"] * n1 + row["pos3p0"] * n2) / n))

				# genomic representative interval should also be updated
				current["start0"] = int(round((current["start0"] * n1 + row["start0"] * n2) / n))
				current["end0"]   = int(round((current["end0"]   * n1 + row["end0"]   * n2) / n))

				current["n_reads"] = n
				current["pos5_list"] += row["pos5_list"]
				current["pos3_list"] += row["pos3_list"]
				current["gstart_list"] += row["gstart_list"]
				current["gend_list"] += row["gend_list"]
				current["read_ids"] += row["read_ids"]
			else:
				out_rows.append(current)
				current = row.copy()

		if current is not None:
			out_rows.append(current)

	out = pd.DataFrame(out_rows)

	# recompute representative values from merged lists for robustness
	out["pos5p0"] = out["pos5_list"].apply(lambda v: int(np.median(np.array(v, dtype=int))))
	out["pos3p0"] = out["pos3_list"].apply(lambda v: int(np.median(np.array(v, dtype=int))))
	out["start0"] = out["gstart_list"].apply(lambda v: int(np.median(np.array(v, dtype=int))))
	out["end0"]   = out["gend_list"].apply(lambda v: int(np.median(np.array(v, dtype=int))))

	out["start_spread_mad_bp"] = out["pos5_list"].apply(lambda v: mad(np.array(v, dtype=int)))
	out["end_spread_mad_bp"]   = out["pos3_list"].apply(lambda v: mad(np.array(v, dtype=int)))

	return out

def summarize_overlaps_for_isoform(chrom: str,
								   strand: str,
								   start0: int,
								   end0: int,
								   genes_chr: pd.DataFrame) -> Dict[str, object]:
	"""
	Annotate one isoform against genes.
	Here start0/end0 MUST be genomic interval bounds with start0 < end0.
	"""
	start0 = int(start0)
	end0   = int(end0)

	if end0 <= start0:
		# return a safe empty annotation instead of raising
		return {
			"isoform_len_bp": 0,
			"sense_overlap_bp": 0,
			"antisense_overlap_bp": 0,
			"intergenic_bp": 0,
			"frac_sense": 0.0,
			"frac_antisense": 0.0,
			"frac_intergenic": 0.0,
			"n_sense_genes": 0,
			"n_antisense_genes": 0,
			"sense_gene_ids": "",
			"sense_gene_names": "",
			"antisense_gene_ids": "",
			"antisense_gene_names": "",
			"class_main": "invalid",
			"class_detail": "invalid_interval",
			"nearest_left_gene_id": "",
			"nearest_left_gene_name": "",
			"nearest_left_dist_bp": np.nan,
			"nearest_right_gene_id": "",
			"nearest_right_gene_name": "",
			"nearest_right_dist_bp": np.nan,
		}

	iso_len = end0 - start0
	cand = genes_chr[(genes_chr["end0"] > start0) & (genes_chr["start0"] < end0)].copy()

	sense_rows = []
	antisense_rows = []
	sense_bp = 0
	antisense_bp = 0

	for _, g in cand.iterrows():
		ol = overlap_len(start0, end0, int(g["start0"]), int(g["end0"]))
		if ol < MIN_GENE_OVERLAP_BP:
			continue

		rec = {
			"gene_id": g["gene_id"],
			"gene_name": g["gene_name"],
			"gene_start0": int(g["start0"]),
			"gene_end0": int(g["end0"]),
			"gene_strand": g["strand"],
			"overlap_bp": int(ol),
			"overlap_frac_isoform": ol / iso_len,
			"overlap_frac_gene": ol / max(1, int(g["end0"]) - int(g["start0"])),
		}

		if g["strand"] == strand:
			sense_rows.append(rec)
			sense_bp += ol
		else:
			antisense_rows.append(rec)
			antisense_bp += ol

	sense_bp = min(sense_bp, iso_len)
	antisense_bp = min(antisense_bp, iso_len)

	# union coverage by any gene to estimate intergenic part
	any_gene_intervals = cand[["start0", "end0"]].sort_values(["start0", "end0"]).to_numpy().tolist()

	merged = []
	for s, e in any_gene_intervals:
		s = max(start0, int(s))
		e = min(end0, int(e))
		if e <= s:
			continue
		if not merged or s > merged[-1][1]:
			merged.append([s, e])
		else:
			merged[-1][1] = max(merged[-1][1], e)

	covered_any_gene_bp = sum(e - s for s, e in merged)
	intergenic_bp = max(0, iso_len - covered_any_gene_bp)

	frac_sense = sense_bp / iso_len
	frac_antisense = antisense_bp / iso_len
	frac_intergenic = intergenic_bp / iso_len

	if frac_intergenic >= PURE_FRAC:
		class_main = "intergenic"
	elif frac_sense >= PURE_FRAC:
		class_main = "sense"
	elif frac_antisense >= PURE_FRAC:
		class_main = "antisense"
	else:
		class_main = "mixed"

	n_sense = len(sense_rows)
	n_antisense = len(antisense_rows)

	if class_main == "intergenic":
		class_detail = "pure_intergenic"
	elif class_main == "sense":
		class_detail = "sense_single_gene" if n_sense == 1 else "sense_multigene"
	elif class_main == "antisense":
		class_detail = "antisense_single_gene" if n_antisense == 1 else "antisense_multigene"
	else:
		if frac_sense > 0 and frac_intergenic > 0 and frac_antisense == 0:
			class_detail = "mixed_sense_intergenic"
		elif frac_antisense > 0 and frac_intergenic > 0 and frac_sense == 0:
			class_detail = "mixed_antisense_intergenic"
		elif frac_sense > 0 and frac_antisense > 0 and frac_intergenic == 0:
			class_detail = "mixed_sense_antisense"
		else:
			class_detail = "mixed_complex"

	left_genes = genes_chr[genes_chr["end0"] <= start0]
	right_genes = genes_chr[genes_chr["start0"] >= end0]

	nearest_left_gene_id = ""
	nearest_left_gene_name = ""
	nearest_left_dist_bp = np.nan
	if not left_genes.empty:
		lg = left_genes.iloc[left_genes["end0"].argmax()]
		nearest_left_gene_id = lg["gene_id"]
		nearest_left_gene_name = lg["gene_name"]
		nearest_left_dist_bp = int(start0 - lg["end0"])

	nearest_right_gene_id = ""
	nearest_right_gene_name = ""
	nearest_right_dist_bp = np.nan
	if not right_genes.empty:
		rg = right_genes.iloc[right_genes["start0"].argmin()]
		nearest_right_gene_id = rg["gene_id"]
		nearest_right_gene_name = rg["gene_name"]
		nearest_right_dist_bp = int(rg["start0"] - end0)

	return {
		"isoform_len_bp": iso_len,
		"sense_overlap_bp": int(sense_bp),
		"antisense_overlap_bp": int(antisense_bp),
		"intergenic_bp": int(intergenic_bp),
		"frac_sense": float(frac_sense),
		"frac_antisense": float(frac_antisense),
		"frac_intergenic": float(frac_intergenic),
		"n_sense_genes": int(n_sense),
		"n_antisense_genes": int(n_antisense),
		"sense_gene_ids": ",".join([x["gene_id"] for x in sense_rows]),
		"sense_gene_names": ",".join([x["gene_name"] for x in sense_rows]),
		"antisense_gene_ids": ",".join([x["gene_id"] for x in antisense_rows]),
		"antisense_gene_names": ",".join([x["gene_name"] for x in antisense_rows]),
		"class_main": class_main,
		"class_detail": class_detail,
		"nearest_left_gene_id": nearest_left_gene_id,
		"nearest_left_gene_name": nearest_left_gene_name,
		"nearest_left_dist_bp": nearest_left_dist_bp,
		"nearest_right_gene_id": nearest_right_gene_id,
		"nearest_right_gene_name": nearest_right_gene_name,
		"nearest_right_dist_bp": nearest_right_dist_bp,
	}


# ----------------------------
# Load genes
# ----------------------------
genes = load_genes_from_gff(GFF_PATH, feature_types=GENE_FEATURE_TYPES)
print(f"Loaded gene annotations: n={len(genes)}")

genes_by_chrom = {
	chrom: g.reset_index(drop=True)
	for chrom, g in genes.groupby("chrom", sort=False)
}


# ----------------------------
# BAM index if needed
# ----------------------------
if not os.path.exists(BAM_PATH + ".bai"):
	pysam.index(BAM_PATH)

bam = pysam.AlignmentFile(BAM_PATH, "rb")


# ----------------------------
# Pass 1: collect reads into isoform bins
# key = (chrom, strand, 5'bin, 3'bin)
# ----------------------------
isoform_groups: Dict[Tuple[str, str, int, int], Dict[str, object]] = {}

for chrom in bam.references:
	try:
		iterator = bam.fetch(chrom)
	except ValueError:
		continue

	for read in iterator:
		if read.is_unmapped:
			continue
		if REQUIRE_PRIMARY and (read.is_secondary or read.is_supplementary):
			continue

		info = aligned_interval_and_ends(read)
		if info is None:
			continue

		strand = info["strand"]
		g_start0 = info["g_start0"]
		g_end0   = info["g_end0"]
		pos5p0   = info["pos5p0"]
		pos3p0   = info["pos3p0"]

		sk = bin_pos(pos5p0, BIN5)
		ek = bin_pos(pos3p0, BIN3)

		key = (chrom, strand, sk, ek)
		if key not in isoform_groups:
			isoform_groups[key] = {
				"chrom": chrom,
				"strand": strand,
				"start_key": sk,
				"end_key": ek,
				"pos5_list": [],
				"pos3_list": [],
				"gstart_list": [],
				"gend_list": [],
				"read_ids": [],
			}

		isoform_groups[key]["pos5_list"].append(int(pos5p0))
		isoform_groups[key]["pos3_list"].append(int(pos3p0))
		isoform_groups[key]["gstart_list"].append(int(g_start0))
		isoform_groups[key]["gend_list"].append(int(g_end0))
		isoform_groups[key]["read_ids"].append(read.query_name)

bam.close()


# ----------------------------
# Convert raw bins to DataFrame
# ----------------------------
rows = []
for (chrom, strand, sk, ek), d in isoform_groups.items():
	n = len(d["read_ids"])
	if n < MIN_ISOFORM_READS:
		continue

	pos5 = np.array(d["pos5_list"], dtype=int)
	pos3 = np.array(d["pos3_list"], dtype=int)
	gs   = np.array(d["gstart_list"], dtype=int)
	ge   = np.array(d["gend_list"], dtype=int)

	rep_pos5p0 = int(np.median(pos5))
	rep_pos3p0 = int(np.median(pos3))
	rep_start0 = int(np.median(gs))
	rep_end0   = int(np.median(ge))

	# safety
	if rep_end0 <= rep_start0:
		continue

	rows.append({
		"chrom": chrom,
		"strand": strand,
		"start_key": sk,
		"end_key": ek,
		"pos5p0": rep_pos5p0,
		"pos3p0": rep_pos3p0,
		"start0": rep_start0,
		"end0": rep_end0,
		"n_reads": n,
		"pos5_list": d["pos5_list"],
		"pos3_list": d["pos3_list"],
		"gstart_list": d["gstart_list"],
		"gend_list": d["gend_list"],
		"read_ids": d["read_ids"],
		"start_spread_mad_bp": mad(pos5),
		"end_spread_mad_bp": mad(pos3),
	})

clusters = pd.DataFrame(rows)
if clusters.empty:
	raise ValueError("No isoforms produced. Check BAM path or clustering thresholds.")

print(f"Raw isoform bins: n={len(clusters)}")


# ----------------------------
# Optional post-merge
# ----------------------------
if DO_POST_MERGE:
	clusters = post_merge_isoforms(clusters, tol5=MERGE_5P_TOL_BP, tol3=MERGE_3P_TOL_BP)
	print(f"After post-merge: n={len(clusters)}")

clusters = clusters.copy().reset_index(drop=True)
clusters["isoform_id"] = [f"ISO_{i+1:06d}" for i in range(len(clusters))]


# ----------------------------
# Annotate isoforms against genes
# ----------------------------
annot_rows = []

for _, r in clusters.iterrows():
	chrom = r["chrom"]
	strand = r["strand"]
	start0 = int(r["start0"])
	end0   = int(r["end0"])

	if end0 <= start0:
		ann = {
			"isoform_len_bp": 0,
			"sense_overlap_bp": 0,
			"antisense_overlap_bp": 0,
			"intergenic_bp": 0,
			"frac_sense": 0.0,
			"frac_antisense": 0.0,
			"frac_intergenic": 0.0,
			"n_sense_genes": 0,
			"n_antisense_genes": 0,
			"sense_gene_ids": "",
			"sense_gene_names": "",
			"antisense_gene_ids": "",
			"antisense_gene_names": "",
			"class_main": "invalid",
			"class_detail": "invalid_interval",
			"nearest_left_gene_id": "",
			"nearest_left_gene_name": "",
			"nearest_left_dist_bp": np.nan,
			"nearest_right_gene_id": "",
			"nearest_right_gene_name": "",
			"nearest_right_dist_bp": np.nan,
		}
	elif chrom not in genes_by_chrom:
		ann = {
			"isoform_len_bp": end0 - start0,
			"sense_overlap_bp": 0,
			"antisense_overlap_bp": 0,
			"intergenic_bp": end0 - start0,
			"frac_sense": 0.0,
			"frac_antisense": 0.0,
			"frac_intergenic": 1.0,
			"n_sense_genes": 0,
			"n_antisense_genes": 0,
			"sense_gene_ids": "",
			"sense_gene_names": "",
			"antisense_gene_ids": "",
			"antisense_gene_names": "",
			"class_main": "intergenic",
			"class_detail": "pure_intergenic",
			"nearest_left_gene_id": "",
			"nearest_left_gene_name": "",
			"nearest_left_dist_bp": np.nan,
			"nearest_right_gene_id": "",
			"nearest_right_gene_name": "",
			"nearest_right_dist_bp": np.nan,
		}
	else:
		ann = summarize_overlaps_for_isoform(
			chrom=chrom,
			strand=strand,
			start0=start0,
			end0=end0,
			genes_chr=genes_by_chrom[chrom]
		)

	rec = {
		"isoform_id": r["isoform_id"],
		"chrom": chrom,
		"strand": strand,

		# genomic interval
		"start0": start0,
		"end0": end0,

		# transcript-oriented ends
		"pos5p0": int(r["pos5p0"]),
		"pos3p0": int(r["pos3p0"]),

		"n_reads": int(r["n_reads"]),
		"start_spread_mad_bp": float(r["start_spread_mad_bp"]),
		"end_spread_mad_bp": float(r["end_spread_mad_bp"]),
		"start_key": int(r["start_key"]),
		"end_key": int(r["end_key"]),
	}
	rec.update(ann)
	annot_rows.append(rec)

isoforms_annot = pd.DataFrame(annot_rows)

isoforms_annot = isoforms_annot.sort_values(
	["chrom", "start0", "end0", "strand", "n_reads"],
	ascending=[True, True, True, True, False]
).reset_index(drop=True)

print("\nIsoform class counts:")
print(isoforms_annot["class_main"].value_counts(dropna=False))


# ----------------------------
# Optional membership table
# ----------------------------
if OUT_MEMBERSHIP_TSV is not None:
	mem_rows = []
	for _, r in clusters.iterrows():
		isoform_id = r["isoform_id"]
		chrom = r["chrom"]
		strand = r["strand"]
		for rid, p5, p3, gs, ge in zip(
			r["read_ids"], r["pos5_list"], r["pos3_list"], r["gstart_list"], r["gend_list"]
		):
			mem_rows.append({
				"isoform_id": isoform_id,
				"chrom": chrom,
				"strand": strand,
				"read_id": rid,
				"pos5p0": int(p5),
				"pos3p0": int(p3),
				"start0": int(gs),
				"end0": int(ge),
			})
	membership = pd.DataFrame(mem_rows)
	membership.to_csv(OUT_MEMBERSHIP_TSV, sep="\t", index=False)
	print(f"\n✅ wrote {OUT_MEMBERSHIP_TSV} | n_rows={len(membership)}")


# ----------------------------
# Write outputs
# ----------------------------
isoforms_annot.to_csv(OUT_ISOFORMS_TSV, sep="\t", index=False)

with pd.ExcelWriter(OUT_ISOFORMS_XLSX, engine="openpyxl") as writer:
	isoforms_annot.to_excel(writer, sheet_name="isoforms", index=False)

	class_summary = (
		isoforms_annot["class_main"]
		.value_counts(dropna=False)
		.rename_axis("class_main")
		.reset_index(name="n_isoforms")
	)
	class_summary.to_excel(writer, sheet_name="class_summary", index=False)

	detail_summary = (
		isoforms_annot["class_detail"]
		.value_counts(dropna=False)
		.rename_axis("class_detail")
		.reset_index(name="n_isoforms")
	)
	detail_summary.to_excel(writer, sheet_name="detail_summary", index=False)

print(f"\n✅ wrote {OUT_ISOFORMS_TSV}")
print(f"✅ wrote {OUT_ISOFORMS_XLSX}")

display(isoforms_annot.head(20))

### Filter out abnormal isoforms

In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv(OUT_ISOFORMS_TSV, sep="\t")

# -----------------------------------
# Derived metrics
# -----------------------------------
df["abnormal_frac"] = df["frac_antisense"] + df["frac_intergenic"]
df["genic_frac"] = df["frac_sense"] + df["frac_antisense"]

# safer length/read summaries if not already numeric
for c in ["n_reads", "isoform_len_bp", "start_spread_mad_bp", "end_spread_mad_bp",
		  "frac_sense", "frac_antisense", "frac_intergenic", "abnormal_frac"]:
	df[c] = pd.to_numeric(df[c], errors="coerce")

# -----------------------------------
# 1) Main class counts
# -----------------------------------
main_counts = (
	df["class_main"]
	.value_counts(dropna=False)
	.rename_axis("class_main")
	.reset_index(name="n_isoforms")
)

display(main_counts)

# -----------------------------------
# 2) Mixed subclass counts
# -----------------------------------
mixed = df[df["class_main"] == "mixed"].copy()

mixed_counts = (
	mixed["class_detail"]
	.value_counts(dropna=False)
	.rename_axis("class_detail")
	.reset_index(name="n_isoforms")
)

print(f"\nMixed subclass counts (n={len(mixed)} isoforms):")
display(mixed_counts)

# -----------------------------------
# 3) Support-weighted summary by class_detail
# -----------------------------------
summary_by_detail = (
	df.groupby(["class_main", "class_detail"], dropna=False)
	  .agg(
		  n_isoforms=("isoform_id", "size"),
		  total_reads=("n_reads", "sum"),
		  median_reads=("n_reads", "median"),
		  mean_reads=("n_reads", "mean"),
		  median_len_bp=("isoform_len_bp", "median"),
		  mean_len_bp=("isoform_len_bp", "mean"),
		  median_frac_sense=("frac_sense", "median"),
		  median_frac_antisense=("frac_antisense", "median"),
		  median_frac_intergenic=("frac_intergenic", "median"),
		  median_abnormal_frac=("abnormal_frac", "median"),
		  median_start_mad=("start_spread_mad_bp", "median"),
		  median_end_mad=("end_spread_mad_bp", "median"),
	  )
	  .reset_index()
	  .sort_values(["class_main", "total_reads"], ascending=[True, False])
)

display(summary_by_detail)

# -----------------------------------
# 4) Mixed-only quantitative summary
# -----------------------------------
mixed_summary = (
	mixed.groupby("class_detail", dropna=False)
		 .agg(
			 n_isoforms=("isoform_id", "size"),
			 total_reads=("n_reads", "sum"),
			 median_reads=("n_reads", "median"),
			 q75_reads=("n_reads", lambda x: np.quantile(x, 0.75)),
			 median_len_bp=("isoform_len_bp", "median"),
			 median_frac_sense=("frac_sense", "median"),
			 median_frac_antisense=("frac_antisense", "median"),
			 median_frac_intergenic=("frac_intergenic", "median"),
			 median_abnormal_frac=("abnormal_frac", "median"),
			 median_start_mad=("start_spread_mad_bp", "median"),
			 median_end_mad=("end_spread_mad_bp", "median"),
		 )
		 .reset_index()
		 .sort_values("total_reads", ascending=False)
)

display(mixed_summary)

,class_main,n_isoforms
0,sense,60488
1,mixed,21451
2,antisense,829
3,intergenic,26



Mixed subclass counts (n=21451 isoforms):


,class_detail,n_isoforms
0,mixed_sense_intergenic,18596
1,mixed_complex,2162
2,mixed_antisense_intergenic,644
3,mixed_sense_antisense,49


,class_main,class_detail,n_isoforms,total_reads,median_reads,mean_reads,median_len_bp,mean_len_bp,median_frac_sense,median_frac_antisense,median_frac_intergenic,median_abnormal_frac,median_start_mad,median_end_mad
0,antisense,antisense_multigene,511,2048,2.0,4.007828,1782.0,1868.933464,0.000000,0.993598,0.007420,1.000000,0.0,1.00
1,antisense,antisense_single_gene,318,1540,2.0,4.842767,1337.0,1402.044025,0.000000,1.000000,0.000000,1.000000,0.0,1.25
2,intergenic,pure_intergenic,26,1235,6.0,47.500000,595.5,641.269231,0.000000,0.000000,1.000000,1.000000,0.0,1.00
6,mixed,mixed_sense_intergenic,18596,927080,4.0,49.853732,1527.0,1683.022317,0.914860,0.000000,0.083538,0.083538,0.5,1.00
3,mixed,mixed_antisense_intergenic,644,35317,3.0,54.840062,1511.5,1654.936335,0.000000,0.874534,0.124155,1.000000,0.0,1.00
4,mixed,mixed_complex,2162,19566,3.0,9.049954,1878.0,2031.677613,0.599780,0.304431,0.069247,0.405803,0.0,1.00
5,mixed,mixed_sense_antisense,49,333,4.0,6.795918,1432.0,1466.244898,0.794059,0.248469,0.000000,0.248469,0.0,1.50
7,sense,sense_multigene,36843,816601,4.0,22.164346,1965.0,2125.560161,0.987613,0.000000,0.012468,0.012470,1.5,1.00
8,sense,sense_single_gene,23645,666395,4.0,28.183337,1286.0,1388.209431,1.000000,0.000000,0.000000,0.000000,1.5,1.00


,class_detail,n_isoforms,total_reads,median_reads,q75_reads,median_len_bp,median_frac_sense,median_frac_antisense,median_frac_intergenic,median_abnormal_frac,median_start_mad,median_end_mad
3,mixed_sense_intergenic,18596,927080,4.0,11.0,1527.0,0.914860,0.000000,0.083538,0.083538,0.5,1.0
0,mixed_antisense_intergenic,644,35317,3.0,6.0,1511.5,0.000000,0.874534,0.124155,1.000000,0.0,1.0
1,mixed_complex,2162,19566,3.0,5.0,1878.0,0.599780,0.304431,0.069247,0.405803,0.0,1.0
2,mixed_sense_antisense,49,333,4.0,7.0,1432.0,0.794059,0.248469,0.000000,0.248469,0.0,1.5


In [ ]:
OUT_FOLDER = MOTHER_FOLDER + "/novel_orf_discovery"

Path(OUT_FOLDER).mkdir(parents=True, exist_ok=True)

OUT_FILTERED_XLSX = OUT_FOLDER + "/isoform_abnormal_filtered.xlsx"

df["abnormal_frac"] = df["frac_antisense"] + df["frac_intergenic"]
df["abnormal_bp"] = df["antisense_overlap_bp"] + df["intergenic_bp"]

core_abnormal = df[
	(df["class_main"].isin(["intergenic", "antisense"])) |
	(df["class_detail"].isin(["mixed_sense_intergenic", "mixed_antisense_intergenic"]))
].copy()

print("core_abnormal:", len(core_abnormal))
display(core_abnormal["class_detail"].value_counts(dropna=False))

core_abnormal_strict = core_abnormal[
	(core_abnormal["n_reads"] >= 10) &
	(core_abnormal["abnormal_frac"] >= 0.20) &
	(core_abnormal["abnormal_bp"] >= 30)
].copy()

print("core_abnormal_strict:", len(core_abnormal_strict))
display(core_abnormal_strict["class_detail"].value_counts(dropna=False))

exploratory_mixed = df[
	(df["class_main"] == "mixed") &
	(df["class_detail"].isin(["mixed_sense_antisense", "mixed_complex"])) &
	(df["n_reads"] >= 10) &
	(df["abnormal_frac"] >= 0.30) &
	(df["abnormal_bp"] >= 60)
].copy()

print("exploratory_mixed:", len(exploratory_mixed))
display(exploratory_mixed["class_detail"].value_counts(dropna=False))

orf_input = pd.concat(
	[core_abnormal_strict, exploratory_mixed],
	ignore_index=True
).drop_duplicates(subset=["isoform_id"])

print("Final ORF-input isoforms:", len(orf_input))
display(orf_input["class_detail"].value_counts(dropna=False))

with pd.ExcelWriter(OUT_FILTERED_XLSX, engine="openpyxl") as writer:
	# df.to_excel(writer, sheet_name="all_isoforms", index=False)
	# core_abnormal.to_excel(writer, sheet_name="core_abnormal_raw", index=False)
	core_abnormal_strict.to_excel(writer, sheet_name="core_abnormal_strict", index=False)
	exploratory_mixed.to_excel(writer, sheet_name="exploratory_mixed", index=False)
	orf_input.to_excel(writer, sheet_name="orf_input", index=False)
	mixed_summary.to_excel(writer, sheet_name="mixed_summary", index=False)
	summary_by_detail.to_excel(writer, sheet_name="detail_summary", index=False)

print("Wrote:", OUT_FILTERED_XLSX)
print("ORF-input isoforms:", len(orf_input))

core_abnormal: 20095


class_detail
mixed_sense_intergenic        18596
mixed_antisense_intergenic      644
antisense_multigene             511
antisense_single_gene           318
pure_intergenic                  26
Name: count, dtype: int64

core_abnormal_strict: 686


class_detail
mixed_sense_intergenic        536
mixed_antisense_intergenic     92
antisense_multigene            26
antisense_single_gene          22
pure_intergenic                10
Name: count, dtype: int64

exploratory_mixed: 130


class_detail
mixed_complex            129
mixed_sense_antisense      1
Name: count, dtype: int64

Final ORF-input isoforms: 816


class_detail
mixed_sense_intergenic        536
mixed_complex                 129
mixed_antisense_intergenic     92
antisense_multigene            26
antisense_single_gene          22
pure_intergenic                10
mixed_sense_antisense           1
Name: count, dtype: int64

Wrote: ../novel_orf_discovery/isoform_abnormal_filtered.xlsx
ORF-input isoforms: 816


In [ ]:
orf_copy = orf_input.copy()

orf_ranked = orf_copy.sort_values(
	["n_reads", "abnormal_bp", "abnormal_frac"],
	ascending=[False, False, False]
)

print("Top ORF-input isoforms:")
columns = ["strand", "start0", "end0", "n_reads", "isoform_len_bp", "frac_sense", 
"frac_antisense", "frac_intergenic", "abnormal_frac", "abnormal_bp", "class_detail", 'sense_gene_ids', 'antisense_gene_ids']

class_details = ["mixed_complex", "mixed_sense_antisense", "mixed_antisense_intergenic"]

display(orf_ranked[orf_ranked["class_detail"].isin(class_details)].head(20)[columns])

Top ORF-input isoforms:


,strand,start0,end0,n_reads,isoform_len_bp,frac_sense,frac_antisense,frac_intergenic,abnormal_frac,abnormal_bp,class_detail,sense_gene_ids,antisense_gene_ids
11,+,27522,28407,15477,885,0.000000,0.749153,0.242938,0.992090,878,mixed_antisense_intergenic,NaN,MMSYN1_0918
12,+,27522,28435,7785,913,0.000000,0.726177,0.266156,0.992333,906,mixed_antisense_intergenic,NaN,MMSYN1_0918
16,+,27522,28630,7537,1108,0.000000,0.598375,0.395307,0.993682,1101,mixed_antisense_intergenic,NaN,MMSYN1_0918
785,-,812150,813731,517,1581,0.192283,0.576850,0.230867,0.807717,1277,mixed_complex,"MMSYN1_0678,MMSYN1_0679,MMSYN1_0680,MMSYN1_0681",MMSYN1_0677
783,-,812150,813533,406,1383,0.109906,0.659436,0.230658,0.890094,1231,mixed_complex,"MMSYN1_0678,MMSYN1_0679",MMSYN1_0677
765,+,441031,442934,337,1903,0.657383,0.267998,0.094587,0.362585,690,mixed_complex,"MMSYN1_0349,MMSYN1_0350,MMSYN1_0352,MMSYN1_0353",MMSYN1_0351
713,+,67152,68661,278,1509,0.560636,0.376408,0.074221,0.450630,680,mixed_complex,MMSYN1_0041,MMSYN1_0040
7,+,27121,28406,260,1285,0.000000,0.833463,0.166537,1.000000,1285,mixed_antisense_intergenic,NaN,"MMSYN1_0917,MMSYN1_0918"
751,+,389566,391154,244,1588,0.318640,0.450882,0.230479,0.681360,1082,mixed_complex,"MMSYN1_0307,MMSYN1_0308",MMSYN1_0306
747,+,375240,376291,223,1051,0.254044,0.489058,0.256898,0.745956,784,mixed_complex,MMSYN1_0294,MMSYN1_0293


## Enumerate & Evaluate Novel ORFs via OSTIR

**Unified pipeline:** For each abnormal isoform, OSTIR scans the full transcript sequence
for all start codons (ATG by default) and predicts TIR at each. We then:

1. Run OSTIR on each transcript (no `start_range` → scans all start codons)
2. For each OSTIR hit, find the downstream in-frame stop codon (table 4: TAA, TAG)
3. Translate to protein using genetic code 4 (TGA = W)
4. Annotate overlap with known genes (sense/antisense/intergenic)
5. Classify initiation mode from TIR + ΔG_mRNA:rRNA

This replaces the old Cell 9 (manual ORF enumeration) + Cell 11 (custom SD scan) entirely.

**Inputs:** `orf_input` (816 abnormal isoforms)  
**Outputs:** `orf_annot` DataFrame with all candidate ORFs + OSTIR scores + overlap annotation

**Anti-SD:** `ACCTCCTTT` (9 nt, Syn1 16S 3′ tail)

**Time Cost:** 6 min for 800 transcripts

In [12]:
# ============================================================
# Unified Cell: OSTIR scan + ORF construction + annotation
# ============================================================

from __future__ import annotations
import os, math, time
from pathlib import Path
from typing import Dict, List, Tuple, Optional

import numpy as np
import pandas as pd
import pysam
from ostir import run_ostir

# ----------------------------
# Inputs
# ----------------------------
FASTA_PATH = MOTHER_FOLDER + "/Genomes_Input/syn1_genome.fasta"
GFF_PATH   = MOTHER_FOLDER + "/Genomes_Input/syn1.genes.gff3"

OUT_FOLDER = MOTHER_FOLDER + "/novel_orf_discovery/candidate_orfs"
Path(OUT_FOLDER).mkdir(parents=True, exist_ok=True)

# Validate orf_input
required_iso_cols = [
	"isoform_id", "chrom", "strand", "start0", "end0", "n_reads",
	"class_main", "class_detail",
	"frac_sense", "frac_antisense", "frac_intergenic",
	"sense_overlap_bp", "antisense_overlap_bp", "intergenic_bp"
]
missing = [c for c in required_iso_cols if c not in orf_input.columns]
if missing:
	raise ValueError(f"orf_input is missing required columns: {missing}")


# ----------------------------
# Genetic code 4 (Mycoplasma: TGA = W)
# ----------------------------
STOP_CODONS = {"TAA", "TAG"}

GENETIC_CODE_4 = {
	"TTT":"F","TTC":"F","TTA":"L","TTG":"L",
	"TCT":"S","TCC":"S","TCA":"S","TCG":"S",
	"TAT":"Y","TAC":"Y","TAA":"*","TAG":"*",
	"TGT":"C","TGC":"C","TGA":"W","TGG":"W",
	"CTT":"L","CTC":"L","CTA":"L","CTG":"L",
	"CCT":"P","CCC":"P","CCA":"P","CCG":"P",
	"CAT":"H","CAC":"H","CAA":"Q","CAG":"Q",
	"CGT":"R","CGC":"R","CGA":"R","CGG":"R",
	"ATT":"I","ATC":"I","ATA":"I","ATG":"M",
	"ACT":"T","ACC":"T","ACA":"T","ACG":"T",
	"AAT":"N","AAC":"N","AAA":"K","AAG":"K",
	"AGT":"S","AGC":"S","AGA":"R","AGG":"R",
	"GTT":"V","GTC":"V","GTA":"V","GTG":"V",
	"GCT":"A","GCC":"A","GCA":"A","GCG":"A",
	"GAT":"D","GAC":"D","GAA":"E","GAG":"E",
	"GGT":"G","GGC":"G","GGA":"G","GGG":"G",
}


# ----------------------------
# OSTIR settings
# ----------------------------
ANTI_SD = 'ACCTCCTTT'
OSTIR_THREADS = 8

# TIR thresholds (from canonical Syn1 gene distribution)
TIR_STRONG_THRESHOLD  = 10000
TIR_MOD_THRESHOLD     = 1000
TIR_WEAK_THRESHOLD    = 100

LEADERLESS_MAX_DIST = 5
DG_MRNA_RRNA_SD_THRESHOLD = -7.2

# ORF length thresholds
MIN_AA_LEN_MAIN  = 24
MIN_AA_LEN_MICRO = 15

# Overlap thresholds
OVERLAP_PURE_FRAC = 0.95


# ----------------------------
# Helpers
# ----------------------------
def revcomp(seq: str) -> str:
	comp = str.maketrans("ACGTNacgtn", "TGCANtgcan")
	return seq.translate(comp)[::-1]

def translate_code4(nt_seq: str) -> str:
	seq = nt_seq.upper()
	usable = (len(seq) // 3) * 3
	seq = seq[:usable]
	return "".join(GENETIC_CODE_4.get(seq[i:i+3], "X") for i in range(0, len(seq), 3))

def load_genes_from_gff(gff_path: str) -> pd.DataFrame:
	if "genes" in globals() and isinstance(globals()["genes"], pd.DataFrame):
		g = globals()["genes"].copy()
		if {"chrom","start0","end0","strand","gene_id","gene_name"}.issubset(set(g.columns)):
			return g

	def parse_attrs(attr_str):
		d = {}
		if pd.isna(attr_str): return d
		for field in str(attr_str).split(";"):
			field = field.strip()
			if "=" in field:
				k, v = field.split("=", 1)
				d[k.strip()] = v.strip()
		return d

	rows = []
	with open(gff_path) as fh:
		for line in fh:
			if not line.strip() or line.startswith("#"): continue
			parts = line.rstrip("\n").split("\t")
			if len(parts) != 9: continue
			chrom, source, feature, start1, end1, score, strand, phase, attrs = parts
			if feature != "gene" or strand not in {"+", "-"}: continue
			attrd = parse_attrs(attrs)
			gene_id = attrd.get("locus_tag", attrd.get("ID", ""))
			gene_name = attrd.get("gene", attrd.get("Name", gene_id))
			rows.append({
				"chrom": chrom, "start0": int(start1)-1, "end0": int(end1),
				"strand": strand, "gene_id": gene_id, "gene_name": gene_name
			})
	return pd.DataFrame(rows).sort_values(["chrom","start0"]).reset_index(drop=True)

def tx_local_to_genomic(iso_start0, iso_end0, strand, local_start0, local_end0):
	if strand == "+":
		return iso_start0 + local_start0, iso_start0 + local_end0
	else:
		return iso_end0 - local_end0, iso_end0 - local_start0

def merged_union_len(intervals):
	if not intervals: return 0
	intervals = sorted(intervals)
	merged = []
	for s, e in intervals:
		if e <= s: continue
		if not merged or s > merged[-1][1]:
			merged.append([s, e])
		else:
			merged[-1][1] = max(merged[-1][1], e)
	return sum(e - s for s, e in merged)

def annotate_orf_overlap(chrom, strand, g0, g1, genes_chr):
	orf_len = int(g1) - int(g0)
	if orf_len <= 0:
		return {
			"orf_bp_in_sense":0,"orf_bp_in_antisense":0,"orf_bp_in_intergenic":0,
			"orf_frac_in_sense":0.0,"orf_frac_in_antisense":0.0,"orf_frac_in_intergenic":0.0,
			"orf_sense_gene_ids":"","orf_sense_gene_names":"",
			"orf_antisense_gene_ids":"","orf_antisense_gene_names":"",
			"orf_n_sense_genes":0,"orf_n_antisense_genes":0,
			"orf_novelty_class":"invalid","orf_abnormal_frac":0.0,"orf_abnormal_bp":0,
		}
	cand = genes_chr[(genes_chr["end0"]>g0)&(genes_chr["start0"]<g1)]
	sense_iv, anti_iv, any_iv = [], [], []
	sense_ids, sense_names, anti_ids, anti_names = [], [], [], []
	for _, g in cand.iterrows():
		s, e = max(g0,int(g["start0"])), min(g1,int(g["end0"]))
		if e<=s: continue
		any_iv.append((s,e))
		if g["strand"]==strand:
			sense_iv.append((s,e)); sense_ids.append(str(g["gene_id"])); sense_names.append(str(g["gene_name"]))
		else:
			anti_iv.append((s,e)); anti_ids.append(str(g["gene_id"])); anti_names.append(str(g["gene_name"]))
	bp_s, bp_a, bp_any = merged_union_len(sense_iv), merged_union_len(anti_iv), merged_union_len(any_iv)
	bp_ig = max(0, orf_len-bp_any)
	abn_bp = bp_a + bp_ig
	abn_frac = abn_bp/orf_len
	if bp_ig/orf_len>=OVERLAP_PURE_FRAC: nov="novel_intergenic_orf"
	elif bp_a/orf_len>=OVERLAP_PURE_FRAC: nov="antisense_orf"
	elif bp_s/orf_len>=OVERLAP_PURE_FRAC: nov="known_sense_orf"
	elif abn_frac>=0.20 and bp_s>0: nov="mixed_boundary_orf"
	else: nov="mixed_complex_orf"
	return {
		"orf_bp_in_sense":int(bp_s),"orf_bp_in_antisense":int(bp_a),"orf_bp_in_intergenic":int(bp_ig),
		"orf_frac_in_sense":bp_s/orf_len,"orf_frac_in_antisense":bp_a/orf_len,"orf_frac_in_intergenic":bp_ig/orf_len,
		"orf_sense_gene_ids":",".join(sorted(set(sense_ids))),"orf_sense_gene_names":",".join(sorted(set(sense_names))),
		"orf_antisense_gene_ids":",".join(sorted(set(anti_ids))),"orf_antisense_gene_names":",".join(sorted(set(anti_names))),
		"orf_n_sense_genes":len(set(sense_ids)),"orf_n_antisense_genes":len(set(anti_ids)),
		"orf_novelty_class":nov,"orf_abnormal_frac":float(abn_frac),"orf_abnormal_bp":int(abn_bp),
	}

def low_complexity_peptide(aa_seq):
	if not aa_seq: return True
	freqs = pd.Series(list(aa_seq.upper())).value_counts(normalize=True)
	return float(freqs.max()) >= 0.60

def classify_initiation_ostir(tir, dg_mrna_rrna, leaderless_dist):
	has_sd = (not math.isnan(dg_mrna_rrna)) and (dg_mrna_rrna < DG_MRNA_RRNA_SD_THRESHOLD)
	is_ll = leaderless_dist <= LEADERLESS_MAX_DIST
	if math.isnan(tir) or tir <= 0:
		if is_ll: return {"initiation_mode":"leaderless","initiation_score":1,"has_sd":False}
		return {"initiation_mode":"none","initiation_score":0,"has_sd":False}
	if is_ll and tir >= TIR_WEAK_THRESHOLD:
		sc = 3 if tir>=TIR_STRONG_THRESHOLD else (2 if tir>=TIR_MOD_THRESHOLD else 1)
		return {"initiation_mode":"leaderless","initiation_score":sc,"has_sd":has_sd}
	if is_ll:
		return {"initiation_mode":"leaderless","initiation_score":1,"has_sd":has_sd}
	if tir>=TIR_STRONG_THRESHOLD and has_sd: return {"initiation_mode":"sd_strong","initiation_score":3,"has_sd":True}
	if tir>=TIR_MOD_THRESHOLD and has_sd: return {"initiation_mode":"sd_moderate","initiation_score":2,"has_sd":True}
	if tir>=TIR_WEAK_THRESHOLD and has_sd: return {"initiation_mode":"sd_weak","initiation_score":1,"has_sd":True}
	if tir>=TIR_MOD_THRESHOLD and not has_sd: return {"initiation_mode":"sd_independent","initiation_score":2,"has_sd":False}
	if tir>=TIR_WEAK_THRESHOLD and not has_sd: return {"initiation_mode":"sd_independent_weak","initiation_score":1,"has_sd":False}
	return {"initiation_mode":"none","initiation_score":0,"has_sd":has_sd}


# ----------------------------
# Load genome + genes
# ----------------------------
fasta = pysam.FastaFile(FASTA_PATH)
genes = load_genes_from_gff(GFF_PATH)
genes_by_chrom = {chrom: g.reset_index(drop=True) for chrom, g in genes.groupby("chrom", sort=False)}
print(f"Loaded {len(genes)} genes from {GFF_PATH}")


# ----------------------------
# Core function: OSTIR scan → ORF construction for one isoform
# ----------------------------
def find_stop_codon(tx_seq: str, start_local0: int) -> Tuple[Optional[int], str]:
	"""
	From a start codon at start_local0, scan in-frame for the first stop codon (table 4).
	Returns (stop_pos_local0, stop_codon) or (None, '') if no stop found.
	"""
	seq = tx_seq.upper()
	j = start_local0 + 3
	while j + 3 <= len(seq):
		codon = seq[j:j+3]
		if codon in STOP_CODONS:
			return j, codon
		j += 3
	return None, ""

def process_one_isoform(iso: pd.Series, fasta_file) -> List[Dict]:
	"""
	Run OSTIR on the full transcript sequence, then for each predicted start site:
	  1. Find downstream in-frame stop codon
	  2. Translate to protein
	  3. Apply length filter
	  4. Annotate overlap + initiation mode
	Returns list of ORF dicts.
	"""
	isoform_id = iso["isoform_id"]
	chrom = str(iso["chrom"])
	strand = str(iso["strand"])
	start0 = int(iso["start0"])
	end0 = int(iso["end0"])

	# Fetch transcript sequence (already oriented 5'→3')
	tx_seq = fasta_file.fetch(chrom, start0, end0).upper()
	if strand == "-":
		tx_seq = revcomp(tx_seq)
	tx_len = len(tx_seq)

	if tx_len < 10:
		return []

	# Run OSTIR: scan ALL start codons in the transcript
	try:
		ostir_hits = run_ostir(tx_seq, aSD=ANTI_SD, threads=OSTIR_THREADS)
	except Exception as e:
		print(f"  OSTIR error for {isoform_id}: {e}")
		return []

	if not ostir_hits:
		return []

	orf_rows = []
	for k, hit in enumerate(ostir_hits):
		start_pos_1based = hit.get("start_position", 0)
		local_start0 = start_pos_1based - 1  # convert to 0-based

		if local_start0 < 0 or local_start0 + 3 > tx_len:
			continue

		tir = hit.get("expression", np.nan)
		start_codon = tx_seq[local_start0:local_start0+3]

		# Find downstream stop codon
		stop_pos, stop_codon = find_stop_codon(tx_seq, local_start0)

		if stop_pos is not None:
			# Complete ORF
			orf_end_local0 = stop_pos + 3  # include stop codon
			nt_seq = tx_seq[local_start0:orf_end_local0]
			aa_seq_full = translate_code4(nt_seq)
			aa_seq = aa_seq_full.rstrip("*")
			if aa_seq:
				aa_seq = "M" + aa_seq[1:]  # initiation always uses fMet
			termination = "complete_stop"
		else:
			# No stop within transcript
			usable_end = local_start0 + ((tx_len - local_start0) // 3) * 3
			if usable_end <= local_start0:
				continue
			orf_end_local0 = usable_end
			nt_seq = tx_seq[local_start0:orf_end_local0]
			aa_seq = translate_code4(nt_seq)
			if aa_seq:
				aa_seq = "M" + aa_seq[1:]  # initiation always uses fMet
			termination = "no_stop_within_isoform"

		aa_len = len(aa_seq)
		if aa_len < MIN_AA_LEN_MICRO:
			continue

		length_tier = "main" if aa_len >= MIN_AA_LEN_MAIN else "micro"

		# Genomic coordinates
		g0, g1 = tx_local_to_genomic(start0, end0, strand, local_start0, orf_end_local0)

		# Overlap annotation
		if chrom in genes_by_chrom:
			ov = annotate_orf_overlap(chrom, strand, g0, g1, genes_by_chrom[chrom])
		else:
			orf_len_bp = g1 - g0
			ov = {
				"orf_bp_in_sense":0,"orf_bp_in_antisense":0,"orf_bp_in_intergenic":int(orf_len_bp),
				"orf_frac_in_sense":0.0,"orf_frac_in_antisense":0.0,"orf_frac_in_intergenic":1.0,
				"orf_sense_gene_ids":"","orf_sense_gene_names":"",
				"orf_antisense_gene_ids":"","orf_antisense_gene_names":"",
				"orf_n_sense_genes":0,"orf_n_antisense_genes":0,
				"orf_novelty_class":"novel_intergenic_orf",
				"orf_abnormal_frac":1.0,"orf_abnormal_bp":int(orf_len_bp),
			}

		# Initiation classification
		dg_mrna_rrna = hit.get("dG_rRNA:mRNA", np.nan)
		init = classify_initiation_ostir(tir, dg_mrna_rrna, leaderless_dist=local_start0)

		orf_rows.append({
			# Isoform info
			"orf_id": f"{isoform_id}__ORF_{k+1:03d}",
			"isoform_id": isoform_id,
			"chrom": chrom,
			"strand": strand,
			"isoform_start0": start0,
			"isoform_end0": end0,
			"isoform_len_bp": end0 - start0,
			"isoform_n_reads": int(iso["n_reads"]),
			"class_main": iso["class_main"],
			"class_detail": iso["class_detail"],
			"isoform_frac_sense": float(iso["frac_sense"]),
			"isoform_frac_antisense": float(iso["frac_antisense"]),
			"isoform_frac_intergenic": float(iso["frac_intergenic"]),
			"isoform_sense_overlap_bp": int(iso["sense_overlap_bp"]),
			"isoform_antisense_overlap_bp": int(iso["antisense_overlap_bp"]),
			"isoform_intergenic_bp": int(iso["intergenic_bp"]),
			"tx_seq": tx_seq,
			"tx_len_bp": tx_len,
			# ORF coordinates
			"orf_local_start0": local_start0,
			"orf_local_end0": orf_end_local0,
			"orf_genomic_start0": int(g0),
			"orf_genomic_end0": int(g1),
			"start_codon": start_codon,
			"stop_codon": stop_codon if stop_codon else "",
			"termination_status": termination,
			"orf_nt_len": len(nt_seq),
			"orf_aa_len": aa_len,
			"length_tier": length_tier,
			"orf_nt_seq": nt_seq,
			"orf_aa_seq": aa_seq,
			"orf_aa_seq_with_stop": aa_seq_full if termination == "complete_stop" else aa_seq,
			# OSTIR results
			"ostir_TIR": tir,
			"ostir_dG_total": hit.get("dG_total", np.nan),
			"ostir_dG_mRNA_rRNA": dg_mrna_rrna,
			"ostir_dG_mRNA": hit.get("dG_mRNA", np.nan),
			"ostir_dG_spacing": hit.get("dG_spacing", np.nan),
			"ostir_dG_standby": hit.get("dG_standby", np.nan),
			"ostir_dG_start_codon": hit.get("dG_start_codon", np.nan),
			"ostir_RBS_distance_bp": hit.get("RBS_distance_bp", np.nan),
			"ostir_start_codon": hit.get("start_codon", ""),
			# Initiation classification
			**init,
			# Overlap + novelty
			**ov,
			# Flags
			"leaderless": local_start0 <= LEADERLESS_MAX_DIST,
			"leaderless_dist_from_5p_nt": local_start0,
			"is_novel_orf": ov["orf_novelty_class"] in [
				"novel_intergenic_orf","antisense_orf","mixed_boundary_orf","mixed_complex_orf"
			],
			"low_complexity_peptide": bool(low_complexity_peptide(aa_seq)),
		})

	return orf_rows


# ----------------------------
# Main loop over all isoforms
# ----------------------------
all_orf_rows = []
t0 = time.time()
n_iso = len(orf_input)

for idx, (_, iso) in enumerate(orf_input.iterrows()):
	if idx % 50 == 0:
		elapsed = time.time() - t0
		print(f'  [{elapsed:6.0f}s]  Isoform {idx+1:>4d}/{n_iso}  '
			  f'({iso["isoform_id"]})  ORFs so far: {len(all_orf_rows)}')

	orf_rows = process_one_isoform(iso, fasta)
	all_orf_rows.extend(orf_rows)

elapsed_total = time.time() - t0

# Build the giant DataFrame
orf_annot = pd.DataFrame(all_orf_rows)

print(f"\nFinished: {n_iso} isoforms in {elapsed_total:.0f}s ({elapsed_total/60:.1f} min)")
print(f"Total candidate ORFs: {len(orf_annot)}")

if len(orf_annot) > 0:
	print(f"\ntermination_status:")
	print(orf_annot["termination_status"].value_counts())
	print(f"\nlength_tier:")
	print(orf_annot["length_tier"].value_counts())
	print(f"\nInitiation mode:")
	print(orf_annot["initiation_mode"].value_counts())
	print(f"\nNovelty class:")
	print(orf_annot["orf_novelty_class"].value_counts())
	print(f"\nNovel ORFs: {orf_annot['is_novel_orf'].sum()}")
	print(f"\nTIR summary (all valid):")
	valid_tir = orf_annot["ostir_TIR"].dropna()
	if len(valid_tir) > 0:
		print(f"  n={len(valid_tir)}, median={valid_tir.median():.1f}, "
			  f"mean={valid_tir.mean():.1f}, range=[{valid_tir.min():.2f}, {valid_tir.max():.0f}]")

	# Save
	OUT_RAW_TSV  = OUT_FOLDER + "/candidate_orfs_ostir.tsv"
	OUT_RAW_XLSX = OUT_FOLDER + "/candidate_orfs_ostir.xlsx"
	orf_annot.to_csv(OUT_RAW_TSV, sep="\t", index=False)
	with pd.ExcelWriter(OUT_RAW_XLSX, engine="openpyxl") as writer:
		orf_annot.to_excel(writer, sheet_name="all_orfs", index=False)
	print(f"\n✅ wrote {OUT_RAW_TSV}")
	print(f"✅ wrote {OUT_RAW_XLSX}")
else:
	print("WARNING: No ORFs found!")

# Also keep as raw_orfs for backward compatibility with downstream cells
raw_orfs = orf_annot

Loaded 911 genes from ../Genomes_Input/syn1.genes.gff3
  [     0s]  Isoform    1/816  (ISO_000189)  ORFs so far: 0


  [    12s]  Isoform   51/816  (ISO_001793)  ORFs so far: 3277


  [    26s]  Isoform  101/816  (ISO_002481)  ORFs so far: 8022


  [    36s]  Isoform  151/816  (ISO_002697)  ORFs so far: 10594


  [    44s]  Isoform  201/816  (ISO_037663)  ORFs so far: 11645


  [    54s]  Isoform  251/816  (ISO_004589)  ORFs so far: 12499


  [    64s]  Isoform  301/816  (ISO_005236)  ORFs so far: 13879


  [    74s]  Isoform  351/816  (ISO_008219)  ORFs so far: 15520


  [    84s]  Isoform  401/816  (ISO_011737)  ORFs so far: 16435


  [    93s]  Isoform  451/816  (ISO_054545)  ORFs so far: 17360


  [   103s]  Isoform  501/816  (ISO_029335)  ORFs so far: 18309


  [   113s]  Isoform  551/816  (ISO_030248)  ORFs so far: 19271
  [   122s]  Isoform  601/816  (ISO_074972)  ORFs so far: 20251


  [   132s]  Isoform  651/816  (ISO_035227)  ORFs so far: 21273


  [   142s]  Isoform  701/816  (ISO_002824)  ORFs so far: 22842


  [   153s]  Isoform  751/816  (ISO_017620)  ORFs so far: 24133


  [   164s]  Isoform  801/816  (ISO_033206)  ORFs so far: 25449



Finished: 816 isoforms in 168s (2.8 min)
Total candidate ORFs: 25991

termination_status:
termination_status
complete_stop             23991
no_stop_within_isoform     2000
Name: count, dtype: int64

length_tier:
length_tier
main     19046
micro     6945
Name: count, dtype: int64

Initiation mode:
initiation_mode
none                   17335
sd_independent_weak     6838
sd_independent          1060
sd_moderate              296
sd_weak                  222
sd_strong                207
leaderless                33
Name: count, dtype: int64

Novelty class:
orf_novelty_class
known_sense_orf         16636
antisense_orf            3751
novel_intergenic_orf     2438
mixed_boundary_orf       1755
mixed_complex_orf        1411
Name: count, dtype: int64

Novel ORFs: 9355

TIR summary (all valid):
  n=25991, median=26.7, mean=648.6, range=[0.00, 227196]

✅ wrote ../novel_orf_discovery/candidate_orfs/candidate_orfs_ostir.tsv
✅ wrote ../novel_orf_discovery/candidate_orfs/candidate_orfs_ostir.xlsx


## Filtering and Deduplication

synthesis_rate = RNA reads × OSTIR TIR

Filters: aa > 15, abnormal_frac > 0.2, not known_sense_orf
Then deduplicate by exact peptide sequence, take top 100 by synthesis rate.

Canonical protein has synthesis rate of 4k read depth times 5k TIR = 20 M.

In [13]:
# ============================================================
# Cell 3
# Synthesis rate ranking → top 100 → then check uniqueness
# ============================================================

from __future__ import annotations
import pandas as pd
import numpy as np

if "orf_annot" not in globals():
	raise ValueError("orf_annot not found. Run the OSTIR cell first.")


# ----------------------------
# Synthesis rate = RNA reads × TIR
# ----------------------------
scored = orf_annot.copy()
scored["ostir_TIR"] = pd.to_numeric(scored["ostir_TIR"], errors="coerce").fillna(0)
scored["isoform_n_reads"] = pd.to_numeric(scored["isoform_n_reads"], errors="coerce").fillna(0)

scored["synthesis_rate"] = scored["isoform_n_reads"] * scored["ostir_TIR"]

print(f"All scored ORFs: {len(scored)}")
print("Synthesis rate summary:")
print(scored["synthesis_rate"].describe().round(1))


# ----------------------------
# Filter
# ----------------------------
filtered = scored[
	(scored["orf_aa_len"] > 15) &
	(scored["orf_abnormal_frac"] > 0.2) &
	(~scored["orf_novelty_class"].isin(["known_sense_orf"]))
].copy()

print(f"\nAfter filtering (aa>15, abnormal>0.2, not known_sense): {len(filtered)}")


# ----------------------------
# Strict set: top 100 by synthesis rate
# ----------------------------
strict_set = (
	filtered
	.sort_values("synthesis_rate", ascending=False)
	.head(100)
	.reset_index(drop=True)
	.copy()
)

strict_set["strict_rank"] = range(1, len(strict_set) + 1)
strict_set["orf_label"] = [f"NOVEL_ORF_{i+1:03d}" for i in range(len(strict_set))]

print(f"\nStrict set: {len(strict_set)} ORFs")
print(f"  Synthesis rate range: {strict_set['synthesis_rate'].min():.0f} – {strict_set['synthesis_rate'].max():.0f}")
print(f"  Median aa length: {strict_set['orf_aa_len'].median():.0f}")
print("  Termination status:")
for val, cnt in strict_set["termination_status"].value_counts().items():
	print(f"    {val}: {cnt}")
print("  Novelty classes:")
for val, cnt in strict_set["orf_novelty_class"].value_counts().items():
	print(f"    {val}: {cnt}")
print("  Initiation modes:")
for val, cnt in strict_set["initiation_mode"].value_counts().head(10).items():
	print(f"    {val}: {cnt}")


# ----------------------------
# Check uniqueness within the top 100
# ----------------------------
n_total = len(strict_set)
n_unique_peptides = strict_set["orf_aa_seq"].nunique()
n_duplicate = n_total - n_unique_peptides

dup_peptides = strict_set[strict_set.duplicated("orf_aa_seq", keep=False)]
dup_groups = dup_peptides.groupby("orf_aa_seq").agg(
	n_occurrences=("orf_id", "size"),
	isoform_ids=("isoform_id", lambda x: ",".join(sorted(set(x)))),
	ranks=("strict_rank", lambda x: ",".join(str(r) for r in sorted(x))),
	max_synthesis_rate=("synthesis_rate", "max"),
).reset_index().sort_values("max_synthesis_rate", ascending=False)

print(f"\n--- Uniqueness check ---")
print(f"  Total ORFs in strict set:    {n_total}")
print(f"  Unique peptide sequences:    {n_unique_peptides}")
print(f"  Duplicate entries:           {n_duplicate}")

if len(dup_groups) > 0:
	print(f"\n  Peptides appearing multiple times: {len(dup_groups)}")
	print(dup_groups[["n_occurrences", "isoform_ids", "ranks", "max_synthesis_rate"]].head(15).to_string(index=False))


# ----------------------------
# FASTA writing (all top 100, including duplicates)
# ----------------------------
def write_fasta(df: pd.DataFrame, fasta_path: str):
	with open(fasta_path, "w") as fh:
		for _, r in df.iterrows():
			header = (
				f">{r['orf_label']}"
				f"|orf_id={r['orf_id']}"
				f"|aa={int(r['orf_aa_len'])}"
				f"|term={r['termination_status']}"
				f"|synth_rate={float(r['synthesis_rate']):.0f}"
				f"|TIR={float(r['ostir_TIR']):.0f}"
				f"|reads={int(r['isoform_n_reads'])}"
				f"|novelty={r['orf_novelty_class']}"
				f"|isoform={r['isoform_id']}"
			)
			fh.write(header + "\n")
			fh.write(str(r["orf_aa_seq"]) + "\n")


# ----------------------------
# Write outputs
# ----------------------------
OUT_SCORED_TSV   = OUT_FOLDER + "/candidate_orfs_scored.tsv"
OUT_STRICT_TSV   = OUT_FOLDER + "/candidate_orfs_strict_top100.tsv"
OUT_STRICT_XLSX  = OUT_FOLDER + "/candidate_orfs_strict_top100.xlsx"
OUT_STRICT_FASTA = OUT_FOLDER + "/novel_orfs_strict.fasta"

scored.to_csv(OUT_SCORED_TSV, sep="\t", index=False)
strict_set.to_csv(OUT_STRICT_TSV, sep="\t", index=False)

with pd.ExcelWriter(OUT_STRICT_XLSX, engine="openpyxl") as writer:
	strict_set.to_excel(writer, sheet_name="strict_top100", index=False)
	if len(dup_groups) > 0:
		dup_groups.to_excel(writer, sheet_name="duplicate_peptides", index=False)

write_fasta(strict_set, OUT_STRICT_FASTA)

print(f"\n✅ wrote {OUT_SCORED_TSV}  (all {len(scored)} scored ORFs)")
print(f"✅ wrote {OUT_STRICT_TSV}")
print(f"✅ wrote {OUT_STRICT_XLSX}")
print(f"✅ wrote {OUT_STRICT_FASTA}  ({len(strict_set)} ORFs, {n_unique_peptides} unique peptides)")

display(strict_set.head(20))


All scored ORFs: 25991
Synthesis rate summary:
count        25991.0
mean         87735.5
std        2809934.5
min              0.0
25%            140.5
50%           1203.6
75%           5904.6
max      258135453.3
Name: synthesis_rate, dtype: float64

After filtering (aa>15, abnormal>0.2, not known_sense): 8296

Strict set: 100 ORFs
  Synthesis rate range: 569786 – 214703029
  Median aa length: 51
  Termination status:
    complete_stop: 91
    no_stop_within_isoform: 9
  Novelty classes:
    novel_intergenic_orf: 64
    mixed_complex_orf: 17
    antisense_orf: 12
    mixed_boundary_orf: 7
  Initiation modes:
    sd_strong: 40
    sd_independent_weak: 31
    none: 12
    sd_independent: 9
    sd_weak: 6
    sd_moderate: 2

--- Uniqueness check ---
  Total ORFs in strict set:    100
  Unique peptide sequences:    44
  Duplicate entries:           56

  Peptides appearing multiple times: 18
 n_occurrences                                                                                   

,orf_id,isoform_id,chrom,strand,isoform_start0,isoform_end0,isoform_len_bp,isoform_n_reads,class_main,class_detail,...,orf_novelty_class,orf_abnormal_frac,orf_abnormal_bp,leaderless,leaderless_dist_from_5p_nt,is_novel_orf,low_complexity_peptide,synthesis_rate,strict_rank,orf_label
0,ISO_034684__ORF_002,ISO_034684,CP002027.1,+,985516,986735,1219,2070,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,0.993827,161,False,23,True,False,2.147030e+08,1,NOVEL_ORF_001
1,ISO_034674__ORF_002,ISO_034674,CP002027.1,+,985516,986444,928,1063,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,0.993827,161,False,23,True,False,1.102557e+08,2,NOVEL_ORF_002
2,ISO_060441__ORF_071,ISO_060441,CP002027.1,-,728029,730604,2575,890,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,1.000000,357,False,1848,True,False,6.597051e+07,3,NOVEL_ORF_003
3,ISO_060447__ORF_071,ISO_060447,CP002027.1,-,728197,730604,2407,813,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,1.000000,357,False,1848,True,False,6.026295e+07,4,NOVEL_ORF_004
4,ISO_060463__ORF_071,ISO_060463,CP002027.1,-,728686,730604,1918,213,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,1.000000,69,False,1848,True,False,1.578845e+07,5,NOVEL_ORF_005
5,ISO_078119__ORF_026,ISO_078119,CP002027.1,-,950610,951741,1131,6513,mixed,mixed_sense_intergenic,...,mixed_boundary_orf,0.208333,40,False,492,True,False,1.570110e+07,6,NOVEL_ORF_006
6,ISO_034675__ORF_002,ISO_034675,CP002027.1,+,985516,986484,968,149,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,0.993827,161,False,23,True,False,1.545447e+07,7,NOVEL_ORF_007
7,ISO_060457__ORF_071,ISO_060457,CP002027.1,-,728504,730604,2100,182,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,1.000000,252,False,1848,True,False,1.349060e+07,8,NOVEL_ORF_008
8,ISO_060329__ORF_002,ISO_060329,CP002027.1,-,728029,728772,743,43,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,1.000000,357,False,16,True,False,9.769417e+06,9,NOVEL_ORF_009
9,ISO_043118__ORF_099,ISO_043118,CP002027.1,-,262395,264676,2281,41,mixed,mixed_sense_intergenic,...,novel_intergenic_orf,1.000000,117,False,2141,True,False,7.748984e+06,10,NOVEL_ORF_010


In [14]:
# ----------------------------
# Collapse strict_set to unique peptides
# ----------------------------
def join_unique(vals) -> str:
	vals = [str(v) for v in vals if str(v) not in {"", "nan", "None"}]
	return ",".join(sorted(set(vals)))

unique_peptides = (
	strict_set
	.groupby("orf_aa_seq", dropna=False)
	.agg(
		n_entries=("orf_id", "size"),
		total_synthesis_rate=("synthesis_rate", "sum"),
		all_isoform_ids=("isoform_id", join_unique),
		n_isoforms=("isoform_id", lambda x: len(set(x))),
		all_orf_ids=("orf_id", join_unique),
		chrom=("chrom", "first"),
		strand=("strand", "first"),
		orf_genomic_start0=("orf_genomic_start0", "first"),
		orf_genomic_end0=("orf_genomic_end0", "first"),
		start_codon=("start_codon", "first"),
		stop_codon=("stop_codon", "first"),
		termination_status=("termination_status", "first"),
		orf_aa_len=("orf_aa_len", "first"),
		orf_nt_seq=("orf_nt_seq", "first"),
		orf_aa_seq_with_stop=("orf_aa_seq_with_stop", "first"),
		novelty_classes=("orf_novelty_class", join_unique),
		initiation_modes=("initiation_mode", join_unique),
	)
	.reset_index()
	.sort_values("total_synthesis_rate", ascending=False)
	.reset_index(drop=True)
)

unique_peptides["peptide_id"] = [f"NOVEL_PEP_{i+1:03d}" for i in range(len(unique_peptides))]

print(f"Unique peptide sequences: {len(unique_peptides)}")
print(f"Total synthesis rate range: {unique_peptides['total_synthesis_rate'].min():.0f} – {unique_peptides['total_synthesis_rate'].max():.0f}")
print(f"Median aa length: {unique_peptides['orf_aa_len'].median():.0f}")

# ----------------------------
# Write unique peptides FASTA
# ----------------------------
OUT_UNIQUE_FASTA = OUT_FOLDER + "/novel_peptides_unique.fasta"

with open(OUT_UNIQUE_FASTA, "w") as fh:
	for _, r in unique_peptides.iterrows():
		header = (
			f">{r['peptide_id']}"
			f"|aa={int(r['orf_aa_len'])}"
			f"|term={r['termination_status']}"
			f"|total_synth_rate={float(r['total_synthesis_rate']):.0f}"
			f"|n_isoforms={int(r['n_isoforms'])}"
			f"|novelty={r['novelty_classes']}"
			f"|isoforms={r['all_isoform_ids']}"
		)
		fh.write(header + "\n")
		fh.write(str(r["orf_aa_seq"]) + "\n")

print(f"\n✅ wrote {OUT_UNIQUE_FASTA}  ({len(unique_peptides)} unique peptides)")

Unique peptide sequences: 44
Total synthesis rate range: 578617 – 350726000
Median aa length: 50

✅ wrote ../novel_orf_discovery/candidate_orfs/novel_peptides_unique.fasta  (44 unique peptides)


## *In silico* trypsin digestion to Find MS Detectable ORF 

In [15]:
from collections import defaultdict

# ----------------------------
# Trypsin digestion settings
# ----------------------------
CANONICAL_FASTA = MOTHER_FOLDER + "/Genomes_Input/syn1_proteins.faa"

OUT_DIGEST_FOLDER = MOTHER_FOLDER + "/novel_orf_discovery/trypsin_digest"
Path(OUT_DIGEST_FOLDER).mkdir(parents=True, exist_ok=True)

MAX_MISSED_CLEAVAGES = 1
MIN_PEPTIDE_LEN = 7
MAX_PEPTIDE_LEN = 25


# ----------------------------
# FASTA reader
# ----------------------------
def read_fasta(path: str):
	header = None
	seq_chunks = []
	with open(path, "r") as fh:
		for line in fh:
			line = line.strip()
			if not line:
				continue
			if line.startswith(">"):
				if header is not None:
					yield header, "".join(seq_chunks)
				header = line[1:]
				seq_chunks = []
			else:
				seq_chunks.append(line)
		if header is not None:
			yield header, "".join(seq_chunks)


# ----------------------------
# Trypsin digestion
# ----------------------------
def trypsin_cleavage_sites(seq: str) -> list:
	seq = seq.upper()
	cuts = [0]
	for i in range(len(seq) - 1):
		if seq[i] in {"K", "R"} and seq[i + 1] != "P":
			cuts.append(i + 1)
	cuts.append(len(seq))
	return cuts

def digest_trypsin(seq: str, max_mc: int = 1) -> list:
	seq = seq.upper()
	cuts = trypsin_cleavage_sites(seq)
	peptides = []
	for i in range(len(cuts) - 1):
		for mc in range(max_mc + 1):
			j = i + mc + 1
			if j >= len(cuts):
				continue
			pep = seq[cuts[i]:cuts[j]]
			if len(pep) == 0 or "X" in pep:
				continue
			peptides.append({
				"peptide_seq": pep,
				"pep_len": len(pep),
				"missed_cleavages": mc,
			})
	return peptides


# ----------------------------
# Digest canonical proteome
# ----------------------------
canonical_peptide_set = set()
for header, seq in read_fasta(CANONICAL_FASTA):
	for p in digest_trypsin(seq, max_mc=MAX_MISSED_CLEAVAGES):
		canonical_peptide_set.add(p["peptide_seq"])

print(f"\nCanonical tryptic peptides: {len(canonical_peptide_set)}")


# ----------------------------
# Digest unique novel peptides
# ----------------------------
novel_tryptic_rows = []
novel_peptide_to_orfs = defaultdict(set)

for _, r in unique_peptides.iterrows():
	pid = r["peptide_id"]
	for p in digest_trypsin(r["orf_aa_seq"], max_mc=MAX_MISSED_CLEAVAGES):
		pep = p["peptide_seq"]
		detectable = MIN_PEPTIDE_LEN <= p["pep_len"] <= MAX_PEPTIDE_LEN
		in_canonical = pep in canonical_peptide_set
		novel_peptide_to_orfs[pep].add(pid)
		novel_tryptic_rows.append({
			"novel_peptide_id": pid,
			"tryptic_peptide": pep,
			"pep_len": p["pep_len"],
			"missed_cleavages": p["missed_cleavages"],
			"detectable_len": detectable,
			"in_canonical": in_canonical,
		})

tryptic_df = pd.DataFrame(novel_tryptic_rows)

# Annotate uniqueness
tryptic_df["unique_vs_canonical"] = ~tryptic_df["in_canonical"]
tryptic_df["n_novel_orfs_sharing"] = tryptic_df["tryptic_peptide"].map(
	lambda x: len(novel_peptide_to_orfs[x])
)
tryptic_df["globally_unique"] = tryptic_df["unique_vs_canonical"] & (tryptic_df["n_novel_orfs_sharing"] == 1)
tryptic_df["usable_for_ms"] = tryptic_df["detectable_len"] & tryptic_df["unique_vs_canonical"]

print(f"Total tryptic peptide rows: {len(tryptic_df)}")
print(f"Unique tryptic sequences: {tryptic_df['tryptic_peptide'].nunique()}")


# ----------------------------
# ORF-level summary
# ----------------------------
orf_digest_summary = tryptic_df.groupby("novel_peptide_id").agg(
	n_tryptic_total=("tryptic_peptide", "size"),
	n_detectable=("detectable_len", "sum"),
	n_unique_vs_canonical=("usable_for_ms", "sum"),
	n_globally_unique=("globally_unique", lambda x: x.sum()),
	example_unique_peptides=("tryptic_peptide", lambda x: ",".join(
		x[tryptic_df.loc[x.index, "usable_for_ms"]].drop_duplicates().head(5).tolist()
	)),
).reset_index()

# Merge back with unique_peptides metadata
orf_digest_summary = unique_peptides.merge(orf_digest_summary, left_on="peptide_id", right_on="novel_peptide_id", how="left")

# MS identifiable = has at least 1 unique vs canonical detectable peptide
ms_identifiable = orf_digest_summary[orf_digest_summary["n_unique_vs_canonical"] >= 1].copy()

print(f"\n--- MS identifiability ---")
print(f"  Total unique novel peptides:    {len(orf_digest_summary)}")
print(f"  MS-identifiable (≥1 unique):    {len(ms_identifiable)}")
print(f"  With ≥2 unique tryptic peps:    {(orf_digest_summary['n_unique_vs_canonical'] >= 2).sum()}")


# ----------------------------
# Final FASTA: MS-identifiable unique peptides
# ----------------------------
OUT_FINAL_FASTA = OUT_DIGEST_FOLDER + "/novel_peptides_ms_identifiable.fasta"

with open(OUT_FINAL_FASTA, "w") as fh:
	for _, r in ms_identifiable.iterrows():
		header = (
			f">{r['peptide_id']}"
			f"|aa={int(r['orf_aa_len'])}"
			f"|term={r['termination_status']}"
			f"|total_synth_rate={float(r['total_synthesis_rate']):.0f}"
			f"|n_isoforms={int(r['n_isoforms'])}"
			f"|n_unique_tryptic={int(r['n_unique_vs_canonical'])}"
			f"|novelty={r['novelty_classes']}"
			f"|isoforms={r['all_isoform_ids']}"
		)
		fh.write(header + "\n")
		fh.write(str(r["orf_aa_seq"]) + "\n")


# ----------------------------
# Write all outputs
# ----------------------------
OUT_DIGEST_XLSX = OUT_DIGEST_FOLDER + "/novel_peptide_trypsin_analysis.xlsx"
OUT_SUMMARY_TSV = OUT_DIGEST_FOLDER + "/novel_peptide_digest_summary.tsv"

orf_digest_summary.to_csv(OUT_SUMMARY_TSV, sep="\t", index=False)

with pd.ExcelWriter(OUT_DIGEST_XLSX, engine="openpyxl") as writer:
	orf_digest_summary.to_excel(writer, sheet_name="digest_summary", index=False)
	ms_identifiable.to_excel(writer, sheet_name="ms_identifiable", index=False)
	tryptic_df.to_excel(writer, sheet_name="all_tryptic_peptides", index=False)

print(f"\n✅ wrote {OUT_UNIQUE_FASTA}  ({len(unique_peptides)} unique peptides)")
print(f"✅ wrote {OUT_FINAL_FASTA}  ({len(ms_identifiable)} MS-identifiable peptides)")
print(f"✅ wrote {OUT_SUMMARY_TSV}")
print(f"✅ wrote {OUT_DIGEST_XLSX}")

display(ms_identifiable.head(20))



Canonical tryptic peptides: 61097
Total tryptic peptide rows: 754
Unique tryptic sequences: 390

--- MS identifiability ---
  Total unique novel peptides:    44
  MS-identifiable (≥1 unique):    43
  With ≥2 unique tryptic peps:    42

✅ wrote ../novel_orf_discovery/candidate_orfs/novel_peptides_unique.fasta  (44 unique peptides)
✅ wrote ../novel_orf_discovery/trypsin_digest/novel_peptides_ms_identifiable.fasta  (43 MS-identifiable peptides)
✅ wrote ../novel_orf_discovery/trypsin_digest/novel_peptide_digest_summary.tsv
✅ wrote ../novel_orf_discovery/trypsin_digest/novel_peptide_trypsin_analysis.xlsx


,orf_aa_seq,n_entries,total_synthesis_rate,all_isoform_ids,n_isoforms,all_orf_ids,chrom,strand,orf_genomic_start0,orf_genomic_end0,...,orf_aa_seq_with_stop,novelty_classes,initiation_modes,peptide_id,novel_peptide_id,n_tryptic_total,n_detectable,n_unique_vs_canonical,n_globally_unique,example_unique_peptides
0,MKSSTNKKATLVCVECLSRNYSVNKSGLTQKQRLEIKKFCTTCNAH...,12,3.507260e+08,"ISO_034579,ISO_034586,ISO_034599,ISO_034603,IS...",12,"ISO_034579__ORF_049,ISO_034586__ORF_049,ISO_03...",CP002027.1,+,985539,985701,...,MKSSTNKKATLVCVECLSRNYSVNKSGLTQKQRLEIKKFCTTCNAH...,novel_intergenic_orf,sd_strong,NOVEL_PEP_001,NOVEL_PEP_001,21,9,9,15,"MKSSTNK,KATLVCVECLSR,ATLVCVECLSR,ATLVCVECLSRNY..."
1,MNTKQDYINKIDNLVLNSELNYDQKNLIISILNKFDDNDINLHNVY...,11,1.482557e+08,"ISO_060327,ISO_060328,ISO_060329,ISO_060342,IS...",11,"ISO_060327__ORF_001,ISO_060328__ORF_002,ISO_06...",CP002027.1,-,728399,728756,...,MNTKQDYINKIDNLVLNSELNYDQKNLIISILNKFDDNDINLHNVY...,novel_intergenic_orf,sd_strong,NOVEL_PEP_002,NOVEL_PEP_002,33,16,16,2,"MNTKQDYINK,QDYINKIDNLVLNSELNYDQK,IDNLVLNSELNYD..."
2,MLKIIDQPIKFLTSWLEIMFVLSQINEIVLETESLSLK,8,3.477593e+07,"ISO_043023,ISO_043024,ISO_043031,ISO_043032,IS...",8,"ISO_043023__ORF_060,ISO_043024__ORF_060,ISO_04...",CP002027.1,-,262418,262535,...,MLKIIDQPIKFLTSWLEIMFVLSQINEIVLETESLSLK*,novel_intergenic_orf,sd_strong,NOVEL_PEP_003,NOVEL_PEP_003,5,2,2,2,"MLKIIDQPIK,IIDQPIK"
3,MKVKLLILLESMKLLHYHQKKNYTQCLLQAYYTHWEKLWLLLTQLL...,2,1.681004e+07,"ISO_078067,ISO_078119",2,"ISO_078067__ORF_026,ISO_078119__ORF_026",CP002027.1,-,951057,951249,...,MKVKLLILLESMKLLHYHQKKNYTQCLLQAYYTHWEKLWLLLTQLL...,mixed_boundary_orf,sd_independent,NOVEL_PEP_004,NOVEL_PEP_004,19,12,12,3,"VKLLILLESMK,LLILLESMK,LLILLESMKLLHYHQK,LLHYHQK..."
4,MNTKQDYINKIDNLVLNSELNYD,1,1.578845e+07,ISO_060463,1,ISO_060463__ORF_071,CP002027.1,-,728687,728756,...,MNTKQDYINKIDNLVLNSELNYD,novel_intergenic_orf,sd_strong,NOVEL_PEP_005,NOVEL_PEP_005,5,3,3,2,"MNTKQDYINK,QDYINKIDNLVLNSELNYD,IDNLVLNSELNYD"
5,MPAHFLVYSSKKSHYFI,3,1.413493e+07,"ISO_001167,ISO_001168,ISO_001174",3,"ISO_001167__ORF_043,ISO_001168__ORF_043,ISO_00...",CP002027.1,+,28318,28372,...,LPAHFLVYSSKKSHYFI*,novel_intergenic_orf,sd_independent_weak,NOVEL_PEP_006,NOVEL_PEP_006,5,2,2,2,"MPAHFLVYSSK,MPAHFLVYSSKK"
6,MNTKQDYINKIDNLVLNSELNYDQKNLIISILNKFDDNDINLHNVY...,1,1.349060e+07,ISO_060457,1,ISO_060457__ORF_071,CP002027.1,-,728504,728756,...,MNTKQDYINKIDNLVLNSELNYDQKNLIISILNKFDDNDINLHNVY...,novel_intergenic_orf,sd_strong,NOVEL_PEP_007,NOVEL_PEP_007,19,11,11,2,"MNTKQDYINK,QDYINKIDNLVLNSELNYDQK,IDNLVLNSELNYD..."
7,MATARALNALSLRWWSFLPRRQSTWRVILLASAKLSRKCGIISQER...,3,1.225060e+07,"ISO_001167,ISO_001168,ISO_001174",3,"ISO_001167__ORF_012,ISO_001168__ORF_012,ISO_00...",CP002027.1,+,27702,28407,...,MATARALNALSLRWWSFLPRRQSTWRVILLASAKLSRKCGIISQER...,mixed_complex_orf,sd_weak,NOVEL_PEP_008,NOVEL_PEP_008,47,27,27,2,"MATARALNALSLR,ALNALSLR,ALNALSLRWWSFLPR,WWSFLPR..."
8,MWAKGQQKARNRKKAALLAFFHRLRPPDEHHKNRRSSQRWRNPTGL,8,7.571046e+06,"ISO_001915,ISO_002292,ISO_002339,ISO_002597,IS...",8,"ISO_001915__ORF_156,ISO_002292__ORF_110,ISO_00...",CP002027.1,+,34747,34888,...,MWAKGQQKARNRKKAALLAFFHRLRPPDEHHKNRRSSQRWRNPTGL*,novel_intergenic_orf,sd_independent_weak,NOVEL_PEP_009,NOVEL_PEP_009,25,7,7,15,"MWAKGQQK,KAALLAFFHR,AALLAFFHR,AALLAFFHRLRPPDEH..."
9,MFLSLKTLYRKRTVIWPFYFYTVKSRLLCKTT,3,5.656735e+06,"ISO_001167,ISO_001168,ISO_001174",3,"ISO_001167__ORF_002,ISO_001168__ORF_002,ISO_00...",CP002027.1,+,27547,27646,...,MFLSLKTLYRKRTVIWPFYFYTVKSRLLCKTT*,mixed_complex_orf,sd_independent_weak,NOVEL_PEP_010,NOVEL_PEP_010,15,4,4,11,"MFLSLKTLYR,RTVIWPFYFYTVK,TVIWPFYFYTVK,TVIWPFYF..."


## Mass Spec

**Our collaborator found no match to the novel ORFs identified by this pipeline.**

**Re-evaluate the SD strength and translation initiation mechanism against the canonical proteins.**

**Correct the Ribosome Binding Site Accessibility**